src/elexmodel/handlers/data/CombinedData.py

```diff
@@ -163,24 +196,77 @@ def _get_unexpected_units(self, aggregates):
+        y = reporting_units[response_variable]
+        qr = QuantileRegressionSolver()
+        qr.fit(x_data.values, y.values, weights=reporting_units['baseline_weights'].values, taus=[0.5])
+        y_hat = qr.predict(x_data.values)
+        residuals =  y.values - y_hat
+        abs_residuals = np.abs(residuals)
+        threshold = abs_residuals.mean() + outlier_z_threshold * abs_residuals.std()
+        return reporting_units[(abs_residuals > threshold).flatten()].copy()
+
+
+    def _get_non_modeled_units(
+        self,
+        reporting_units,
+        percent_reporting_threshold,
+        turnout_factor_lower,
+        turnout_factor_upper,
+        margin_change_threshold,
+        unit_blocklist,
+        postal_code_blocklist,
+        fit_margin_outlier_model,
+        fit_turnout_outlier_model,
+        outlier_z_threshold
+    ):
+        units_blocklisted = self.data[
+            (self.data["geographic_unit_fips"].isin(unit_blocklist))
+            | (self.data["postal_code"].isin(postal_code_blocklist))
+        ].copy()
+        units_blocklisted["unit_category"] = "non-modeled: blocklisted"
+
         zero_baseline_units = self._get_units_with_baseline_of_zero()
+        units_with_zero_baseline = self.data[self.data["geographic_unit_fips"].isin(zero_baseline_units)].copy()
+        units_with_zero_baseline["unit_category"] = "non-modeled: zero baseline"
 
         units_with_strange_turnout_factor = (
-            self.data[
-                (self.data.percent_expected_vote >= percent_reporting_threshold)
-                & (self.data["geographic_unit_fips"].isin(expected_geographic_units))
-                & (
-                    (self.data["geographic_unit_fips"].isin(zero_baseline_units))
-                    | (self.data.turnout_factor <= turnout_factor_lower)
-                    | (self.data.turnout_factor >= turnout_factor_upper)
-                )
+            reporting_units[ # these are all already reporting and expected units
+                (reporting_units.turnout_factor <= turnout_factor_lower) | (reporting_units.turnout_factor >= turnout_factor_upper)
             ]
-            .drop_duplicates(subset="geographic_unit_fips")
-            .copy()
+        ).copy()
+        units_with_strange_turnout_factor["unit_category"] = "non-modeled: strange turnout factor"
+
+        non_modeled_units_list = [units_blocklisted, units_with_zero_baseline, units_with_strange_turnout_factor]
+
+        if fit_turnout_outlier_model:
+            units_with_strange_turnout_factor_modeled = self._fit_outlier_detection_model(reporting_units, 'turnout_factor', outlier_z_threshold)
+            units_with_strange_turnout_factor_modeled["unit_category"] = "non-modeled: strange turnout factor modeled"
+            non_modeled_units_list.append(units_with_strange_turnout_factor_modeled)
+
+
+        if "margin" in self.estimands:
+            # reporting_units["normalized_margin_change"] = (
```

Comment:
Do you want to keep this commented-out code? 🤔 

Also can you remove the extra blank line above `if "margin" in self.estimands`?  I am genuinely surprised `pre-commit` didn't remove that lol 🤔

Reply:
huh yeah, will remove thanks!

Reply:
done